src/libsyntax/parse/parser.rs

```diff
@@ -3342,7 +3362,15 @@ impl<'a> Parser<'a> {
             return self.parse_while_let_expr(opt_label, span_lo, attrs);
         }
         let cond = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
-        let (iattrs, body) = self.parse_inner_attrs_and_block()?;
+        let (iattrs, body) = self.parse_inner_attrs_and_block()
+            .map_err(|mut err| {
+                if self.is_identifier_not(&cond) {
+                    let msg = "try replacing identifier `not` with the negation operator";
+                    // FIXME: replaced span doesn't include trailing whitespace
```

Comment:
You could merge the `cond.span` with the next span, get the snippet and synthesize a span that _does_ include the whitespace (in a similar way as suggested in https://github.com/rust-lang/rust/pull/47574#issuecomment-367080392).